auth0/auth0-deploy-cli
# Directory JSON actions secrets matches YAML

### 🔧 Changes


The line of code which handles parsing of secrets during export on YAML, was copied to the `context/directory/handlers/actions.ts`, in `function mapToAction(filePath, action) `.  This code preserves the secret placeholder when it detects a type of string, such as `"@@MY_SECRETS@@"`.  If it is an array, it will call `mapSecrets()`.  This is what I believe gets called when the keywords are set to their values defined in the config file during an import to Auth0.

### 📚 References



- Somewhat related issue for the YAML format actions secrets keyword preservation: issue #902 
- Issue #688

### 🔬 Testing


1. Begin with a tenant that contains an Auth0 action.
2. Export from a tenant using the `--format directory`.  
3. View the directory output, and manually edit the json file for this action to have the `"secrets": []` array use a placeholder value instead.  `"secrets": "@@MY_SECRETS@@"`
4. Define the placeholder preservation in your config file used for your export, simple example below
`exports/webapp/myconfig.json`
```
{
  "AUTH0_ALLOW_DELETE": true,
  "AUTH0_PRESERVE_KEYWORDS": true,
  "AUTH0_KEYWORD_REPLACE_MAPPINGS": {
    "MY_SECRETS": [
      {
        "name": "M_TO_M_APP_DOMAIN",
        "value": "##AUTH0_DOMAIN##"
      }
    ]
  }
}
```
5. Next, perform the export again with this config file
```
a0deploy export \
  --strip \
  --format directory \
  --output_folder ./exports/webapp \
  --config_file ./exports/webapp/myconfig.json \
  --debug true
```
6. This error should no longer occur, the export should complete successfully, the secrets placeholder remains
```
2024-09-06T21:15:03.444Z - debug: TypeError: secrets.map is not a function
    at mapSecrets (/Users/sebastiandetering/.nvm/versions/node/v20.12.2/lib/node_modules/auth0-deploy-cli/lib/context/directory/handlers/actions.js:49:24)
    at mapToAction (/Users/sebastiandetering/.nvm/versions/node/v20.12.2/lib/node_modules/auth0-deploy-cli/lib/context/directory/handlers/actions.js:73:18)
```

### 📝 Checklist
